File: sudoku-common/common-web/src/main/java/com/zumin/sudoku/common/web/template/GetStatisticsDataTemplate.java
package com.zumin.sudoku.common.web.template;

import com.zumin.sudoku.common.web.domain.LocalDateRange;
import com.zumin.sudoku.common.web.domain.StatisticsDateRange;
import com.zumin.sudoku.common.web.enums.StatisticsDate;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;

/**
 * 获取统计数据的模板方法
 *
 * @param <T> 统计数据的类型
 */
@AllArgsConstructor
public class GetStatisticsDataTemplate<T> {

  private final StatisticsDateRange dateRange;

  /**
   * 获取统计数据
   *
   * @param callback 获取统计数据的回调方法
   * @return 统计数据列表
   */
  public List<T> getData(StatisticsDataCallback<T> callback) {
    StatisticsDate statisticsDate = dateRange.getStatisticsDate();
    LocalDateRange range = dateRange.getDateRange();
    LocalDate nowDate = statisticsDate.getFirst(range.getStart()), lastDate = statisticsDate.getFirst(range.getEnd());
    List<T> list = new ArrayList<>();
    while (nowDate.compareTo(lastDate) < 0) {
      LocalDate nextDate = statisticsDate.next(nowDate);
      list.add(callback.getData(nowDate, nextDate));
      nowDate = nextDate;
    }
    return list;
  }

  /**
   * 获取统计数据的回调方法
   *
   * @param <T> 统计数据的类型
   */
  public interface StatisticsDataCallback<T> {

    /**
     * 获取[startDate,endDate)中的统计数据
     *
     * @param startDate 开始日期
     * @param endDate   结束日期
     * @return 统计数据
     */
    T getData(LocalDate startDate, LocalDate endDate);
  }

}

File: sudoku-common/common-web/src/main/java/com/zumin/sudoku/common/web/statistics/GetStatisticsDataTemplate.kt
package com.zumin.sudoku.common.web.statistics

import com.zumin.sudoku.common.web.domain.StatisticsDateRange
import java.time.LocalDate

/**
 * 获取统计数据的模板方法
 *
 * @param <T> 统计数据的类型
 */
class GetStatisticsDataTemplate<T>(private val dateRange: StatisticsDateRange) {
  /**
   * 获取统计数据
   *
   * @param callback 获取统计数据的回调方法
   * @return 统计数据列表
   */
  fun getData(callback: (startDate: LocalDate, endDate: LocalDate) -> T): List<T> {
    val statisticsDate = dateRange.statisticsDate
    val (start, end) = dateRange.dateRange
    var nowDate = statisticsDate.getFirst(start!!)
    val lastDate = statisticsDate.getFirst(end!!)
    val result = mutableListOf<T>()
    while (nowDate < lastDate) {
      val nextDate = statisticsDate.next(nowDate)
      result.add(callback.invoke(nowDate, nextDate))
      nowDate = nextDate
    }
    return result
  }
}